File: lib/simple_event_modeling/cart/queries/cart_items_read.rb
# frozen_string_literal: true

module SimpleEventModeling
  module Cart
    module Queries
      # Query object for projecting the current state of a cart from its event stream.
      #
      # Iterates over all cart-related events and builds a projection of items, quantities, and totals.
      #
      # @example
      #   query = Queries::CartItemsRead.new(cart_id, store)
      #   result = query.execute
      #   # => { cart: { cart_id: ..., items: {...}, totals: {...} } }
      #
      # @param aggregate_id [String] The cart's aggregate ID.
      # @param store [EventStore] The event store containing cart events.
      # @return [Hash] The projected cart state.
      class CartItemsRead
        attr_accessor :aggregate_id, :store, :projection

        def initialize(aggregate_id, store)
          @aggregate_id = aggregate_id
          @store = store
          @projection = {}
        end

        def execute
          events = store.get_stream(aggregate_id)
          events.each do |event|
            on(event)
          end
          { cart: @projection }
        end

        def on(event)
          case event
          when DomainEvents::CartCreated
            on_cart_created(event)
          when DomainEvents::ItemAdded
            on_add_item(event)
          when DomainEvents::ItemRemoved
            on_item_removed(event)
          when DomainEvents::CartCleared
            on_cart_cleared(event)
          else
            raise "Unhandled event type: #{event.class}"
          end
        end

        def on_cart_created(event)
          @projection[:cart_id] = event.aggregate_id
          @projection[:items] = {}
          @projection[:totals] = {
            total: 0.0
          }
        end

        def on_add_item(event)
          @projection[:items] ||= {}
          item_id = event.data[:item]
          @projection[:items][item_id] ||= { quantity: 0 }
          @projection[:items][item_id][:quantity] += 1
        end

        def on_item_removed(event)
          item_id = event.data[:item]
          return unless projection[:items].has_key?(item_id)

          projection[:items][item_id][:quantity] -= 1

          projection[:items].delete(item_id) if projection[:items][item_id][:quantity] < 1
        end

        def on_cart_cleared(_event)
          @projection[:items] = {}
        end
      end
    end
  end
end

File: go/gpt5/cart/queries/cart_items_read.go
package queries

import (
	"fmt"

	"gpt5/common"
)

// CartItemsRead projects the current state of a cart from its event stream.
// Result shape mirrors the Ruby version: { cart: { cart_id, items, totals } }
// items is map[itemID] -> { quantity: int }

type CartItemsRead struct {
	aggregateID string
	store       *common.EventStore
	projection  map[string]any
}

func NewCartItemsRead(aggregateID string, store *common.EventStore) *CartItemsRead {
	return &CartItemsRead{
		aggregateID: aggregateID,
		store:       store,
		projection:  map[string]any{},
	}
}

func (q *CartItemsRead) Execute() (map[string]any, error) {
	events, err := q.store.GetStream(q.aggregateID)
	if err != nil {
		return nil, err
	}
	for _, e := range events {
		if err := q.on(e); err != nil {
			return nil, err
		}
	}
	return map[string]any{"cart": q.projection}, nil
}

func (q *CartItemsRead) on(e common.Event) error {
	switch e.Type {
	case "CartCreated":
		return q.onCartCreated(e)
	case "ItemAdded":
		return q.onItemAdded(e)
	case "ItemRemoved":
		return q.onItemRemoved(e)
	case "CartCleared":
		return q.onCartCleared(e)
	default:
		return fmt.Errorf("unhandled event type: %s", e.Type)
	}
}

func (q *CartItemsRead) onCartCreated(e common.Event) error {
	q.projection["cart_id"] = e.AggregateID
	q.projection["items"] = map[string]map[string]int{}
	q.projection["totals"] = map[string]float64{"total": 0.0}
	return nil
}

func (q *CartItemsRead) onItemAdded(e common.Event) error {
	items, _ := q.projection["items"].(map[string]map[string]int)
	if items == nil {
		items = map[string]map[string]int{}
		q.projection["items"] = items
	}
	itemID, _ := e.Data["item"].(string)
	if _, ok := items[itemID]; !ok {
		items[itemID] = map[string]int{"quantity": 0}
	}
	items[itemID]["quantity"]++
	return nil
}

func (q *CartItemsRead) onItemRemoved(e common.Event) error {
	items, _ := q.projection["items"].(map[string]map[string]int)
	if items == nil {
		return nil
	}
	itemID, _ := e.Data["item"].(string)
	if _, ok := items[itemID]; !ok {
		return nil
	}
	items[itemID]["quantity"]--
	if items[itemID]["quantity"] < 1 {
		delete(items, itemID)
	}
	return nil
}

func (q *CartItemsRead) onCartCleared(_ common.Event) error {
	q.projection["items"] = map[string]map[string]int{}
	return nil
}
